src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/ModelAttributes.cs

```diff
@@ -10,60 +10,80 @@
+            TypeAttributes = Attributes;
+        }
 
-            return parameterAttributes.Concat(typeAttributes);
+        /// <summary>
+        /// Creates a new <see cref="ModelAttributes"/> for a property.
+        /// </summary>
+        /// <param name="propertyAttributes">The set of attributes for the property.</param>
+        /// <param name="typeAttributes">
+        /// The set of attributes for the property's <see cref="Type"/>. See <see cref="PropertyInfo.PropertyType"/>.
+        /// </param>
+        public ModelAttributes([NotNull] IEnumerable<object> propertyAttributes, [NotNull] IEnumerable<object> typeAttributes)
+        {
+            PropertyAttributes = propertyAttributes.ToArray();
+            TypeAttributes = typeAttributes.ToArray();
+            Attributes = PropertyAttributes.Concat(TypeAttributes).ToArray();
         }
 
         /// <summary>
+        /// Gets the set of all attributes. If this instance represents the attributes for a property, the attributes
+        /// on the property definition are before those on the property's <see cref="Type"/>.
+        /// </summary>
+        public IReadOnlyList<object> Attributes { get; }
+
+        /// <summary>
+        /// Gets the set of attributes on the property, or <c>null</c> if this instance represents the attributes
+        /// for a <see cref="Type"/>.
+        /// </summary>
+        public IReadOnlyList<object> PropertyAttributes { get; }
+
+        /// <summary>
+        /// Gets the set of attributes on the <see cref="Type"/>. If this instance represents a property,
+        /// then <see cref="TypeAttributes"/> contains attributes retrieved from
+        /// <see cref="PropertyInfo.PropertyType"/>.
+        /// </summary>
+        public IReadOnlyList<object> TypeAttributes { get; }
+
+        /// <summary>
         /// Gets the attributes for the given <paramref name="property"/>.
         /// </summary>
         /// <param name="type">The <see cref="Type"/> in which caller found <paramref name="property"/>.
         /// </param>
         /// <param name="property">A <see cref="PropertyInfo"/> for which attributes need to be resolved.
         /// </param>
-        /// <returns>An <see cref="IEnumerable{object}"/> containing the attributes on the
-        /// <paramref name="property"/> before the attributes on the <paramref name="property"/> type.</returns>
-        public static IEnumerable<object> GetAttributesForProperty([NotNull] Type type, [NotNull] PropertyInfo property)
+        /// <returns>A <see cref="ModelAttributes"/> instance with the attributes of the property.</returns>
+        public static ModelAttributes GetAttributesForProperty([NotNull] Type type, [NotNull] PropertyInfo property)
         {
-            // Return the property attributes first.
             var propertyAttributes = property.GetCustomAttributes();
             var typeAttributes = property.PropertyType.GetTypeInfo().GetCustomAttributes();
```

Comment:
(might have come up when @kirthik originally added `ModelMetadataTypeAttribute` support but) should this handle `ModelMetadataTypeAttribute` on `property.PropertyType`?  e.g. could call `GetAttributesForType(property.PropertyType)`